### ISSUE TITLE
Rolling update of collector can prevent the DNS to be updated in loadbalancingexporter

### ISSUE THREAD

**Issue by ZhengHe-MD** [open-telemetry/opentelemetry-collector-contrib#1690]
**Describe the bug**

Following description is extracted from discussions [1](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1678#issuecomment-732790261) [2](https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1678#issuecomment-732638083):

The following pods are returned for the "otelcol" hostname: ["otelcol-abc", "otelcol-def", "otelcol-ghi"]. This represents a deployment "otelcol" with three replicas
the load balancer is configured with the hostname set to "otelcol.svc.cluster.local", backed by the pods from the deployment "otelcol"
the deployment is updated, causing new pods to appear: ["otelcol-abc", "otelcol-def", "otelcol-ghi", "otelcol-jkl", "otelcol-mno", "otelcol-pqr"].
once the new pods are in the running state, the old pods are removed: ["otelcol-jkl", "otelcol-mno", "otelcol-pqr"]
The problem is that the list of backend servers is outdated after the last step above, causing spans to be sent to "otelcol-abc". Because they are failing, the read lock will prevent the DNS to be updated. Given that the retry is per batch, and not per exporter, we have a scenario where the list of backends won't be updated at all.

**Steps to reproduce**

1. start otelcol as agent, let's call it otel-agent, use dns resolution in loadbalancing exporter
2. start an script continuously reporting spans to otel-agent
3. change the dns config, which resolve to unreachable ip addresses (now the otel-agent stops periodically checking dns updates)
4. change the dns config back won't change anything

**What did you expect to see?**

dns resolves to new ip addresses, and resolves back to original ones, works fine. spans can still be reported to collector as usual.

**What did you see instead?**

as stated above

**What version did you use?**

Version: any, I would say the latest one 0bf52151c029f7c37485ff240bedf94f04b31c96

**What config did you use?**

**Environment**
OS: MacOS
Compiler(if manually compiled): go 1.14

**Additional context**

A quick solution would be only acquiring read lock for the following lines:

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0bf52151c029f7c37485ff240bedf94f04b31c96/exporter/loadbalancingexporter/exporter.go#L209-L210

This case might be yet another argument for recommending and use batching (sending_queue) by default for the underlying exporters, as proposed by @bogdandrutu :

https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/1680/files#r528803681

through which the `consumeTrace` call can return immediately. 

however, we should keep critical area as small as possible, and have not remote calls inside, definitely.

**Comment by ZhengHe-MD** [comment]
@jpkrohling see if there is any additional information is needed?

**Comment by ZhengHe-MD** [comment]
@jpkrohling if there is no better way, I can create a PR with the solution suggested above.

**Comment by jpkrohling** [comment]
From @Vemmy124:

> There is a point, which I would certainly mention. With queue_retry setting the management of retries comes to the certain OTLP exporter. Thus the traces are lost if the backend is changed

Thinking about this problem some more, I think the ideal solution is to return an error from our ConsumeTraces, based on the error we receive from the underlying ConsumeTraces. We should then also add `exporterhelper.TimeoutSettings`, `exporterhelper.QueueSettings` and `exporterhelper.RetrySettings` to the load balancer exporter.

This would allow people to still have a per-backend queue/retry/timeout configuration, while also having the chance of recovering from errors caused by this scenario here.

The one caveat I see is that we potentially make multiple calls to underlying ConsumeTraces, one per trace. So, we'd have to make a decision about what to do when one of the traces fail but not the others: should we swallow the errors, perhaps recording a metric/log entry, or should we return the error for the first batch in error? The side effect is that previously successful traces would be processed again. Perhaps this behavior should be controller via a config option as well?

**Comment by ZhengHe-MD** [comment]
If we do async export, there is nothing the lb-exporter can do about the traces data, unless we have other means of communicating with the underlying exporter. If we do synchronous export, we have to put the queuing and retry logic in lb-exporter, which will take the right ip address after next resolution, but will certainly destroy current design.

But in our case, our systems in production environment have peak hours, and also non-peak hours, we can limit the rolling update time to only in non-peak hours, when there are probably less interesting traces. So we can tolerate a little bit of data loss. However, if a rolling update will stop agents from resolving dns to new ip addresses, that would be certainly bad because we have to restart hundreds of agents.

**Comment by jpkrohling** [comment]
> If we do synchronous export, we have to put the queuing and retry logic in lb-exporter, which will take the right ip address after next resolution, but that will certainly destroy current design.

Not really: there's already a timeout/retry facility that we can try to make use of: https://github.com/open-telemetry/opentelemetry-collector/blob/618095055f4514cbf3066b39ac63281e2f95de9d/exporter/otlpexporter/factory.go#L68-L75

> However, if a rolling update will stop agents from resolving dns to new ip addresses

This shouldn't happen. Are you thinking about using the load balancer as an agent? I see the load balancer as an extra layer in your deployment architecture:

client (your app) -> agent -> load balancer -> collector

**Comment by ZhengHe-MD** [comment]
@jpkrohling 

> Not really: there's already a timeout/retry facility that we can try to make use of: https://github.com/open-telemetry/opentelemetry-collector/blob/618095055f4514cbf3066b39ac63281e2f95de9d/exporter/otlpexporter/factory.go#L68-L75

yes, you're right, we don't need to implement retry/timeout logic in the case of synchronous export.

> Are you thinking about using the load balancer as an agent? I see the load balancer as an extra layer in your deployment architecture: client (your app) -> agent -> load balancer -> collector

yes. I replace jaeger-agent by otel-collector with loadbalancing exporter enabled. Is there anything bad with this setting? or Is there anything good with using two layers of collectors, with the first layer doing loadbalancing export while the second tail-based sampling.

**Comment by jpkrohling** [comment]
The biggest downside I see is very close to the concern you mentioned: ensuring a consistent cluster view across load balancers is critical in getting all the traceIDs sent to the correct collectors. The more load balancers you have, the higher the chances that the cluster view is going to be different. 

I thought about having scalable load balancers (ie, with consistent cluster views using some consensus algorithm) for future versions, and had in mind implementing a new resolver with etcd as the backend, as it would make it possible for the load balancer to react quicker to backend changes.

**Comment by pkositsyn** [comment]
If you ask me, I also don't understand the need in extra layer. It just produces more traffic. You can try to compare your offered configuration to the following:
1. Agent: batch -> load balancing
2. Collectors: *do any work where the full trace isn't needed* -> groupbytrace -> *do some stuff with full traces* -> batch -> your favourite exporter

**Comment by pkositsyn** [comment]
Yes, need to mention, that with static resolver this seems certainly better to me

**Comment by pkositsyn** [comment]
And anyway, having only one instance of loadbalancer isn't fault tolerant at all. Anyway there is a need in synchronization with many instances

**Comment by ZhengHe-MD** [comment]
@pkositsyn I'm also concerned about extra traffic, and what I've done in production is pretty close to what you suggested, with some additional customized processors to spot interesting spans, and a tail-sampling processor to actually do the sampling.

**Comment by ZhengHe-MD** [comment]
@jpkrohling see, now I can understand why you want have an extra layer of collectors in your mind. the only concern I have is that this can make the solution a bit heavier.

**Comment by jpkrohling** [comment]
> having only one instance of loadbalancer isn't fault tolerant at all

Yes: https://github.com/open-telemetry/opentelemetry-collector/issues/1724#issuecomment-691904210
I think you are the users we were looking for :-)

What would you rather use as the source of information? Zookeeper, etcd, something else? Or gossip between the existing instances?

**Comment by ZhengHe-MD** [comment]
I don't know what data consistency level the DNS service supply, let's say it is eventual consistency, then how to build an extra layer of strong consistency when you can receive different versions of data out of order. If it is strong consistency, then it's the same with what etcd or Zookeeper can supply.

**Comment by jpkrohling** [comment]
Sorry, by consistency I mean at the level of the load balancers. The problem in having multiple instances of the load balancer is that each one of them will do a DNS query at their own discretion, meaning that one DNS changes might take up to almost 2x the time interval to be reflected upon all load balancer instances.

Ensuring a consistent view across the cluster means that all clients have to receive the same change and agree on a time to make the change effective. In general terms, we want something like a raft consensus algorithm, but not sure we want to implement it ourselves, as we'd then need to have the load balancer instances aware of all other instances (or at least leaders, depending on the implementation). And there's still the risk of a cluster split brain ;-)

If we use `etcd`, I *think* we can have it notify the consumers when the data source changes, which might be good enough for our use case. The same might exist for Zookeeper. But again, this also requires another piece of infra to install and maintain, as I'm not sure we can use the etcd that comes with Kubernetes...

**Comment by ZhengHe-MD** [comment]
@jpkrohling 
if we have a consensus algorithm built inside load balancer instances, that's not gonna work because what we want is a consistent view over all instances, not a quorum. and what we really want is simultaneously switching to new set of ip addresses, which is just like a two-generals problem.
if we use `etcd`, by notifying the consumers, it is indeed better than polling a consistent dns service, but not as much theoretically, because there are still network requests between collectors and `etcd`, so we can't ensure a simultaneous switch. I would say if everything goes well, it is faster than current solution. But let's think it deeper, if you implement this successfully, continue with the above example, when dns resolution result changes from `["otelcol-abc", "otelcol-def", "otelcol-ghi"]` to `["otelcol-abc", "otelcol-def", "otelcol-ghi", "otelcol-jkl", "otelcol-mno", "otelcol-pqr"]`, the problem this issue presents still exists, with consistent view.

**Comment by pkositsyn** [comment]
There is theoretically no way to guarantee 100% of complete traces via only lb patches. Given that spans from one trace can be split into batches (at least because they can come from different nodes), one batch can be sent to collector instance, then the backend disappears forever and there is nowhere to send second batch to keep the trace complete. To solve this general problem batches themselves should form a quorum on backends (i.e. 1 batch sent to different collectors) and this seems to be an overkill. Provided that we solve the problem only with rolling update, then the state must be transitioned, otherwise the abovementioned problem still exists. Thus it is not lb problem alone.

**Comment by jpkrohling** [comment]
> if we have a consensus algorithm built inside load balancer instances, that's not gonna work because what we want is a consistent view over all instances, not a quorum

The consensus is mostly to elect a leader, which will then establish what's the "right" view of the cluster and coordinate the changes with the other nodes. Without leader(s), we have no way of knowing which view is the correct one.

> what we really want is simultaneously switching to new set of ip addresses

We can't ever guarantee that. Even with the most perfect solution, clocks could still be out of sync, making changes happen at apparently the same clock time and yet, not being simultaneous.

> unless collectors know what is a rolling update transition state, and wait until it finishes

Sounds like we might want a Kubernetes resolver ;-)

> Given that spans from one trace can be split into batches (at least because they can come from different nodes), one batch can be sent to collector instance, then the backend disappear and there is nowhere to send second batch to keep the trace complete

True, and there's nothing we can do here in this case.

> Thus it is not lb problem alone.

The problems we have here are typical problems for stateful distributed systems. The good news here is that tracing data typically doesn't require the same guarantees as other stateful distributed systems, like replicated databases: people can safely assume  that traces might be dropped by their sampling strategy if they are incomplete, as it's very likely that a similar trace was already sent in the past, or will be sent in the future. Missing a few traces here and there isn't a huge problem, generally speaking. It's bad when we are having a rolling update *and* having a highly interesting and unique trace being dropped, but the chances of it happening are quite small.

That said, users can reduce the risk of this happening by adjusting the time interval where DNS queries will happen, at the expense of network calls. Coupled with the new sending queue mechanism that we discussed before, chances are high that only a very small number of the traces will get affected, as new traces that still belong to a failing backend will eventually fail over to a new backend when the ring is updated.

If that still isn't enough for your specific use cases, we can certainly consider implementing other resolvers that will get data updated even faster, but we'll never reach a level where can guarantee that deviations won't happen.

**Comment by ZhengHe-MD** [comment]
> The consensus is mostly to elect a leader, which will then establish what's the "right" view of the cluster and coordinate the changes with the other nodes. Without leader(s), we have no way of knowing which view is the correct one.
> We can't ever guarantee that. Even with the most perfect solution, clocks could still be out of sync, making changes happen at apparently the same clock time and yet, not being simultaneous.

Yes, totally agree. and what I said "we", it is really representing your proposal, which suggest to maintain a consistent view of backends among load balancers. I think this is an overkill, and my issue is not really about that. I also certainly agree that 

> The good news here is that tracing data typically doesn't require the same guarantees as other stateful distributed systems, like replicated databases: people can safely assume that traces might be dropped by their sampling strategy if they are incomplete, as it's very likely that a similar trace was already sent in the past, or will be sent in the future. Missing a few traces here and there isn't a huge problem, generally speaking. It's bad when we are having a rolling update and having a highly interesting and unique trace being dropped, but the chances of it happening are quite small.

and this is exactly what I do:

> That said, users can reduce the risk of this happening by adjusting the time interval where DNS queries will happen, at the expense of network calls. Coupled with the new sending queue mechanism that we discussed before, chances are high that only a very small number of the traces will get affected, as new traces that still belong to a failing backend will eventually fail over to a new backend when the ring is updated.

that's enough for my specific use case. and let's repeat the issue title, "rolling update of collector can stop the periodical check of DNS updates in loadbalancing exporter", I never request data consistency, current implementation will stop the periodical check. and I believe narrow down the critical area is enough for now, and if we do that, we should provide that information in loadbalancing exporter's README.

**Comment by jpkrohling** [comment]
> and let's repeat the issue title, "rolling update of collector can stop the periodical check of DNS updates in loadbalancing exporter"

Sorry, I guess we sidetracked this issue quite a bit :-) To recap: this issue can be addressed by using the sending queue that we discussed earlier, right? There's still a question about what to do when only some items from the batch fail, and I think I would return "success" but record a metric with the number of traces/spans that were "silently" dropped.

**Comment by ZhengHe-MD** [comment]
@jpkrohling 
this issue can be simply addressed by move the read lock acquiring logic around these two lines

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/0bf52151c029f7c37485ff240bedf94f04b31c96/exporter/loadbalancingexporter/exporter.go#L209-L210

and the periodical check will keep going. this is at least what we can do. but that may cause the traces/spans being dropped "silently" as you stated. If sending_queue is enabled, and if I don't take it wrong, the loadbalancing exporter is not able to know anything about whether the batch succeeded, failed or partly for both, because the result doesn't come out immediately. If sending_queue is disabled, the loadbalancing exporter just wait synchronously, but from the codes below

https://github.com/open-telemetry/opentelemetry-collector/blob/6f60275760aea4c0dbc0e494d8dbdac4b8ecf05c/exporter/exporterhelper/tracehelper.go#L68

I believe the number of traces/spans dropped is not returned from `ConsumeTraces`, but this info may be recorded by the underlying exporter.

**Comment by jpkrohling** [comment]
> If sending_queue is enabled, and if I don't take it wrong, the loadbalancing exporter is not able to know anything about whether the batch succeeded

True, but note that this is about the sending queue for the underlying OTLP exporter. People using the sending queue for the load balancer itself aren't affected.

> but this info may be recorded by the underlying exporter.

True. The best we can do is to record a data point whenever we receive an error from the underlying exporter.

**Comment by ZhengHe-MD** [comment]
> True, but note that this is about the sending queue for the underlying OTLP exporter. People using the sending queue for the load balancer itself aren't affected.

yes, if it is the sending queue for load balancer, then it's possible to record more detailed metrics. this is a great point! maybe we should always disable the sending_queue in the underlying OTLP exporter, and make use of it in the load balancer exporter. I get your idea.

**Comment by ZhengHe-MD** [comment]
@jpkrohling

It's been several days, so, how about firstly narrow down the critical area, fix this issue first, and then introduce queuing settings to loadbalancing exporter, with more fine-grained metrics, and enforce the underlying OTLP exporter to disable queuing. If we consent on this, I'd be glad to create PR for this.

**Comment by jpkrohling** [comment]
I'm OK with that!

**Comment by jpkrohling** [comment]
I believe this was addressed by the linked PR. Closing.
